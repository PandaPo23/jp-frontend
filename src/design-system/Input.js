import React from 'react';

import system from '../utils/System';
import Box from './Box';

// import { themeGet } from 'styled-system';

const StyledInput = system(
  'Input',
  {
    extend: Box,
    as: 'input',
    type: 'text',
    boxShadow: 0,
    borderRadius: 'sm',
    bg: 'surface',
    flex: 1,
    border: 1,
    p: 2,
    width: 1,
    value: '',
  },
  (props) => ({
    appearance: 'none',
    // ':focus': {
    //   border: props.border && themeGet(`borders.${props.border}`)(props),
    // },
    // '::placeholder': {
    //   color: themeGet(`colors.misc.muted`)(props),
    // },
    // '::ms-clear': {
    //   display: 'none',
    // },
  })
);

export const Input = React.forwardRef(
  ({ autoFocus = false, onChange, ...props }, ref) => (
    <StyledInput
      ref={ref}
      autoFocus={autoFocus}
      onChange={onChange ? (event) => onChange(event.target.value) : undefined}
      {...props}
    />
  )
);

export default Input;
