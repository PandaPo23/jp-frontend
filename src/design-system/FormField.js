import PropTypes from 'prop-types';
import React from 'react';
import Box from './Box';
import Flex from './Flex';
import Icon from './Icon';
import Label from './Label';
import Input from './Input';
import Select from './Select';
import { keyframes } from '@emotion/core';

const fadeIn = keyframes`
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
`;
const labelStyles = {
  animation: fadeIn + ' 0.3s',
};

const noop = () => {};

const isFormElement = () => {};

export class FormField extends React.Component {
  static defaultProps = {
    // for backwards-compatibility
    onChange: noop,
    homeSearch: false,
    boxShadow: false,
  };
  static formElements = [Input, Select];

  isFormElement = (element) => FormField.formElements.includes(element);
  // for backwards-compatibility
  handleChange = (onChange) => (e) => {
    this.props.onChange(e);
    if (typeof onChange !== 'function') return;
    onChange(e);
  };
  hasValue = () => {
    const { children } = this.props;
    return React.Children.toArray(children).reduce(
      (a, child) =>
        a || (child && isFormElement(child.type) && child.props.value),
      false
    );
  };
  render() {
    const {
      label,
      icon,
      children,
      onChange,
      homeSearch,
      boxShadow,
      ...props
    } = this.props;
    let FieldChild;
    let position = -1;
    let LabelChild;
    let BeforeIcon;
    let AfterIcon;
    let fieldId;
    let fieldPlaceholder;
    React.Children.forEach(children, (child, index) => {
      if (!child) return;
      const { type, props } = child;
      switch (type) {
        case Label:
          LabelChild = child;
          break;
        case Input:
        case Select:
          position = index;
          FieldChild = child;
          fieldId = props.id;
          // For aria-label when Label child is not rendered
          fieldPlaceholder = props.placeholder;
          break;
        case Icon:
          if (position < 0) {
            BeforeIcon = child;
          } else {
            AfterIcon = child;
          }
          break;
        default:
          break;
      }
    });
    const showLabel =
      LabelChild && LabelChild.props.hidden
        ? false
        : this.props.alwaysShowLabel || (LabelChild && this.hasValue());
    return (
      <Box>
        <Flex alignItems="center" width={1} mt={0}>
          {showLabel &&
            React.cloneElement(LabelChild, {
              pl: 1,
              style: labelStyles,
              htmlFor: fieldId,
            })}
          <Flex width="100%" alignItems="center">
            {BeforeIcon && (
              <Flex mr={-1} mt={showLabel ? 0 : 1} px={3} alignItems="centar">
                {BeforeIcon}
              </Flex>
            )}
            {React.cloneElement(FieldChild, {
              'aria-label':
                !showLabel && fieldPlaceholder ? fieldPlaceholder : null,
              boxShadow: boxShadow ? 1 : null,
              border: 0,
              transition: 'padding-top 0.1s, padding-bottom 0.1s',
              width: 1,
              ml: showLabel && BeforeIcon ? -5 : 0,
              mt: showLabel && -2,
              pl: BeforeIcon ? (showLabel ? 5 : 0) : 3,
              pr: AfterIcon && 2,
              pt: showLabel ? 3 : 2,
              pb: 2,
              // ref: (elem) => {
              //   this.fieldRef = elem;
              // },
              // for backwards compatibility
              onChange: this.handleChange(FieldChild.props.onChange),
              ...props,
            })}
          </Flex>
          {AfterIcon && (
            <Box
              alignItems="center"
              ml={-1}
              mt={showLabel ? -2 : 0}
              bg={AfterIcon.props.bg}
              style={{
                height: homeSearch ? '4rem' : null,
                width: homeSearch ? '5rem' : 0,
                display: 'flex',
                alignItems: 'center',
                justifyContent: 'center',
                borderBottomRightRadius: 5,
                borderTopRightRadius: 5,
              }}
            >
              {AfterIcon}
            </Box>
          )}
        </Flex>
      </Box>
    );
  }
}

FormField.propTypes = {
  alwaysShowLabel: PropTypes.bool,
  homeSearch: PropTypes.bool,
  children: PropTypes.any,
};

export default FormField;
