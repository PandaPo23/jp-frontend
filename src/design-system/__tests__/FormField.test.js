import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import FormField from '../FormField';
import Input from '../Input';

describe('FormField component', () => {
  const getComponent = (props) => (
    <FormField>
      <Input {...props} py={3} is="input" type="input" />
    </FormField>
  );
  it('should render as FormField correctly', () => {
    const component = getComponent({
      id: 'idFiled',
      name: 'fieldName',
    });
    const wrapper = mount(component);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
