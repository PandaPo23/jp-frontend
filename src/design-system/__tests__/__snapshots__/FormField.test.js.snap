// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField component should render as FormField correctly 1`] = `
.emotion-2 {
  margin-top: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  background-color: surface;
  margin-left: 0;
  padding: 8px;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 16px;
  width: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: 0;
  border-radius: sm;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

<FormField
  boxShadow={false}
  homeSearch={false}
  onChange={[Function]}
>
  <Box>
    <div
      className="emotion-3"
    >
      <Flex
        alignItems="center"
        display="flex"
        mt={0}
        width={1}
      >
        <div
          className="emotion-2"
        >
          <Flex
            alignItems="center"
            display="flex"
            width="100%"
          >
            <div
              className="emotion-1"
            >
              <ForwardRef
                aria-label={null}
                border={0}
                boxShadow={null}
                id="idFiled"
                is="input"
                ml={0}
                name="fieldName"
                onChange={[Function]}
                pb={2}
                pl={3}
                pt={2}
                py={3}
                transition="padding-top 0.1s, padding-bottom 0.1s"
                type="input"
                width={1}
              >
                <Input
                  aria-label={null}
                  as="input"
                  autoFocus={false}
                  bg="surface"
                  border={0}
                  borderRadius="sm"
                  boxShadow={null}
                  flex={1}
                  id="idFiled"
                  is="input"
                  ml={0}
                  name="fieldName"
                  onChange={[Function]}
                  p={2}
                  pb={2}
                  pl={3}
                  pt={2}
                  py={3}
                  transition="padding-top 0.1s, padding-bottom 0.1s"
                  type="input"
                  value=""
                  width={1}
                >
                  <input
                    aria-label={null}
                    autoFocus={false}
                    className="emotion-0"
                    id="idFiled"
                    name="fieldName"
                    onChange={[Function]}
                    type="input"
                    value=""
                  />
                </Input>
              </ForwardRef>
            </div>
          </Flex>
        </div>
      </Flex>
    </div>
  </Box>
</FormField>
`;
